### PR TITLE
Clear up import policy

### LINKED ISSUE #2468: Clear up imports policy
We're currently not very consistent regarding what we make available directly via `telegram.__init__` and what we don't. Some examples:

* from `telegram.error` only `TelegramError` is available via `telegram.__init__`, but no of the other classes
* from `telegram.constants` only some few constants are available.
* `TelegramDeprecationWarning` is not available (see also #2467)
* the helper methods from `utils.helper` that are actually there to be used by users and not exclusively by PTB itself (e.g. `escape_markdown`, `mention_user_*`) are not available

IMO we should try to make a clear policy for this. IMHO it should be clear that all API-classes are directly available through `telgeram`. All the things mentioned above are from … lets call them "auxiliary modules". For these I basically see two options:

1. Make all of that available directly through `telegram`
2. Make none of it available through `telegram`. In this case I'd vote to at least move the above mentioned helper methods directly to `telegram.utils` and spare the user the `.helpers`, especially as users are in general not to be concerned with the other contents of `utils`.

Case 2 would be breaking & in that case we should wait until v14.
In any case I would aim to document the import policy somewhere … probably just at the index of the docs.
I'm slightly in favor of 2. What do the other @python-telegram-bot/developers think?

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
closes #2468

Some notes:

- I know 120 files sounds bad the the majority of changes is really just imports :D 
- `ext.utils.promise.Promise` was deliberately not moved to `ext.promise.Promise` because since the removal of `MessageQueue` users should almost never have contact with that class. You can view that at [RTD](https://python-telegram-bot.readthedocs.io/en/clear-up-imports/telegram.html#auxiliary-modules)
- While at it I split up that messy `tg.utils.helpers` module into a number of separate modules and moved parts of the contents just to the place where it's actually used. Also added a bunch of additional tests
- Added a new section in the docs for the modules in `telegram` that are not strictly API related

TODO:

- [x] check if we can add some more `versionchanged` directives
- [x] Wait for #2662 to me merged so that we can include the changes to the warnings
- [x] There's now `tg.helpers` and `tg.utils.helpers` maybe we can find a better name for one of the two to avoid confusion …

### Checklist for PRs

- [x] Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)  -- This can probably still be improved

### CODE REVIEW

**Review by harshil21 — CHANGES_REQUESTED** [review]
Good changes. Some things we can still change:

- While at it, we can also rename base.py to telegramobject.py, cause for some reason in rtd the 'bases' section doesn't link is broken (just updated my sphinx to 4.2.0 from 4.0.2 and now this is fixed!), and I suppose telegramobject.py is more explicit.

- Also `telegram.ext.filters Module` is listed under the ['Handlers' section](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.html#handlers). Maybe move it up where the rest of telegram.ext is (i.e with Updater and the rest)?

- I noticed `telegramobject.rst` doesn't have `:github_url:`, maybe add that?

File: telegram/ext/commandhandler.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ class CommandHandler(Handler[Update, CCT]):
         attributes to :class:`telegram.ext.CallbackContext`. See its docs for more info.
 
     Args:
-        command (:class:`telegram.utils.types.SLT[str]`):
+        command (:obj:`str` | Tuple[:obj:`str`] | List[:obj:`str`]):
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/commandhandler.py:56]: because of that and because `utils.types` is not part of the public API. don't know why I agreed to adding it in the first place :D

File: telegram/ext/updater.py [hunk 2]
```diff
@@ -38,19 +38,29 @@
...
+from telegram.utils.defaultvalue import DEFAULT_FALSE, DefaultValue
+from telegram.utils.warnings import warn
 from telegram.ext.utils.types import CCT, UD, CD, BD
 from telegram.ext.utils.webhookhandler import WebhookAppClass, WebhookServer
 
 if TYPE_CHECKING:
     from telegram.ext import BasePersistence, Defaults, CallbackContext
 
 
+# From https://stackoverflow.com/questions/2549939/get-signal-names-from-numbers-in-python
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/updater.py:55]: I was unsure, because that's not documented and because of the comment about pylint. But it seems that it's indeed just not documented, but it does exist:

* https://github.com/python/typeshed/pull/555#issuecomment-247874222
* https://bugs.python.org/issue28206

There was a [PR](https://github.com/python/cpython/pull/1939) which never made it for some reason - maybe I'll pick it up 😄 

TL;DR: will change

File: telegram/message.py [hunk 3]
```diff
@@ -17,7 +17,13 @@
 #
 # You should have received a copy of the GNU Lesser Public License
 # along with this program.  If not, see [http://www.gnu.org/licenses/].
-"""This module contains an object that represents a Telegram Message."""
+"""This module contains an object that represents a Telegram Message.
+
+Warning:
+    Contents of this module are intended to be used internally by the library and *not* by the
+    user. Changes to this module are not considered breaking changes and may not be documented in
+    the changelog.
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/message.py:25]: image

Me trying to figure out why I did that

**Comment by Bibo-Joshi** [comment]
> * While at it, we can also rename base.py to telegramobject.py, cause for some reason in rtd the 'bases' section doesn't link is broken (just updated my sphinx to 4.2.0 from 4.0.2 and now this is fixed!), and I suppose telegramobject.py is more explicit.

The broken link is not really related to the file name, but I renamed anyway. see sphinx-doc/sphinx#9026
But this is actually a good reminder to follow up on sphinx-doc/sphinx#4944 being fixed now 👍 we should do that in the doc-fixes branch though

> * Also `telegram.ext.filters Module` is listed under the ['Handlers' section](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.html#handlers). Maybe move it up where the rest of telegram.ext is (i.e with Updater and the rest)?

I moved it there on purpose a while back, because `filters` basically belongs to `MessageHandler` … at least for me that makes more sense …

> * I noticed `telegramobject.rst` doesn't have `:github_url:`, maybe add that?

✔️

File: telegram/utils/warnings.py [hunk 4]
```diff
@@ -61,8 +38,10 @@ def warn(message: str, category: Type[Warning] = PTBUserWarning, stacklevel: int
     .. versionadded:: 14.0
 
     Args:
-        category (:obj:`Type[Warning]`): Specify the Warning class to pass to ``warnings.warn()``.
-        stacklevel (:obj:`int`): Specify the stacklevel to pass to ``warnings.warn()``. Pass the
-            same value as you'd pass directly to ``warnings.warn()``.
+        message (:obj:`str`): Specify the warnings message to pass to ``warnings.warn()``.
+        category (:obj:`Type[Warning]`, optional): Specify the Warning class to pass to
+            ``warnings.warn()``. Defaults to :class:`telegram.warnings.PTBUserWarning``
```
Comment (harshil21) [inline telegram/utils/warnings.py:43]:
```suggestion
            ``warnings.warn()``. Defaults to :class:`telegram.warnings.PTBUserWarning`
```
